packages/opencensus-core/src/trace/instrumentation/base-plugin.ts

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as shimmer from 'shimmer';
+import * as types from './types';
+import * as modelTypes from '../model/types';
+
+/** This class represent the base to patch plugin. */
+export abstract class BasePlugin implements types.Plugin {
+
+  /** The service to send the collected traces */
+  // tslint:disable:no-any
+  protected exporter: any;
+  /** The module name */
+  protected moduleName: string;
+  /** A tracer object. */
+  protected tracer: modelTypes.Tracer;
+  /** The module version. */
+  protected version: string;
+
+  /**
+   * Constructs a new BasePlugin instance.
+   * @param moduleName The module name.
+   */
+  constructor(moduleName: string) {
+    this.moduleName = moduleName;
+  }
+
+  /**
+   * Sets modified plugin to the context.
+   * @param exporter object module to set on context
+   * @param tracer tracer relating to context
+   * @param version module version description
+   */
+  // tslint:disable:no-any
+  protected setPluginContext(exporter: any, tracer: modelTypes.Tracer, version: string) {
+    this.exporter = exporter;
+    this.tracer = tracer;
+    this.version = version;
+  }
+
+  /**
+   * Wraps a function.
+   * @param nodule The module.
+   * @param name The function name.
+   * @param wrapper The wrapper.
+   */
+  protected wrap(nodule, name, wrapper) {
```

Comment:
I agree the implementations should be free to choose between wrapping functions, but the existence of this functions in class BasePlugin does not make them mandatory to be used by implementations. One advantage is that it removes shimmer dependency from implementations.

Reply:
It's OK for instrumentations to depend on `shimmer` since npm will install a single module for all of them anyway, so there is no dependency size penalty. I would still prefer to keep the API as minimal as possible, especially functions like these that are little more than call-throughs to an underlying dependency module.